### DISCUSSION TITLE
Configuring authentication

### DISCUSSION THREAD

**Discussion by necik11** [pnp/pnpcore#1105] [Q&A] [answered]
I would like to ask how to properly configure authentication.. and which type to use "best practices". I would like to use the application (it will run, for example, on my server/azure) and I will connect to other tenants (multiple).

The 1st step is clear: app registration + API permissions (use application permissions or delegate permissions? )
I know i can Use Delegate permissions with UsernNamePassword auth but when I choose certificate should not i use app permissions?

 step number 2 authentication?

there is many approaches:
https://pnp.github.io/pnpcore/using-the-sdk/configuring%20authentication.html?q=authentication
![auth](https://user-images.githubusercontent.com/25322303/219979237-edb512fa-cc17-40b0-9dc2-984186ad25c3.PNG)

InteractiveOptions - when the app is running in the background I can't use the ...

UsernamePasswordAuthenticationProvider - here I have to solve where to store the password? but what if there are multiple tenants I want to connect to? so add them all to the credential manager? see:
https://pnp.github.io/pnpcore/using-the-sdk/configuring%20authentication.html#using-the-credential-manager

or best choice is use PnPCoreAuthenticationX509CertificateOptions and foreach instance/app make certificate?

Thanks for tips :)

**Comment by jansenbe** [comment]
@necik11 : if your code is going to run on an Azure backend than you've two choices. If the code needs to run unattended (so not being directly called from a user context) then you should use Application permissions, which in case of PnP Core SDK requires the X509CertificateAuthentication provider. When your code on the Azure server is being called from a client and you want your server code to act as the client that called it then you need to use the OnBehalfOfAuthentication provider

**Comment by necik11** [comment, marked as answer]
Okay i will try better describe:

- I wanna have application at cloud(azure) in app services / but for testing i would rather use my desktop. 

-  For sample console app and there is:

- ![image](https://user-images.githubusercontent.com/25322303/220293023-6199f5af-a3df-4061-860f-5f25c2ebca18.png)

1. . Cause i wanna make service job - i cant use **interactive auth**...
2. **UserNamepassword auth** - this work but i want open more tenants - for now i have all logins credentials in json... And i am considering save passwords to keyvaults...? But i read i should use Client secrets or better way Certificates and this keys save to keyvaults?... 

- There is no choice for fill clientSecret in userNamePassword - I think cause i should use onBehalf? 

4. When i choose **OnBehalf auth** its want userToken - so i must first send post and get Token? I didnt find any tutorial... only testing scenarios

- ![image](https://user-images.githubusercontent.com/25322303/220299359-b4d4d229-41c2-4549-8fbd-69f026cb7cea.png)

5. Last i was trying **X509Certificate auth**

- From this [manual](https://learn.microsoft.com/en-us/azure/active-directory/develop/howto-create-self-signed-certificate)

And there is:
If your application will be running from another machine or cloud, such as Azure Automation, you'll also need a private key.

Export-Pfx ... but if you go app registrations, there is limit only .cer, .pem or .crt so certificate pfx with private key cant be uploaded
![certAzure1](https://user-images.githubusercontent.com/25322303/220303011-004619dc-110f-415c-83ba-6e5687ddb298.PNG)

- When i try :
- Private key Certificate

` X509CertificateAuthenticationProvider authProvider = new X509CertificateAuthenticationProvider(clientID, tenantID, cert);`

I got error - {"The certificate certificate does not have a private key. "} - But how i can upload when app registration not allow? 

I found only when i publish code into App Service... but i wanna 1 App whichi will join on Multiple Tenant.. and at each tenant i just wanna allow permision for some account... so for this usage i need usage Delegate option and just login and password? 

But for this i need disable Multi factor auth... ?

- Public Certificate with thumbprint

I got Error: {"error_description":"ID3035: The request was not valid or is malformed."}

6. I even find this manual for configure [certificate ](https://learn.microsoft.com/en-us/azure/app-service/configure-ssl-certificate?tabs=apex%2Cportal) but this is just for 1 application .. like i wrote many times i need have 1 app which will connect on Many tenants and do some stuffs :)

So i need know which type authentication(login+pass/cert/clientSecret/onbehalf) in my console app AND which type permission (App X Delegate) on each app registristation:
![AzureApp1](https://user-images.githubusercontent.com/25322303/220307310-a8db5a4b-0280-437f-be13-5cc9924082bd.png)


Thanks much for any advance - i can give more pictures or better describes if its needed ? :) 👍

**Reply by jansenbe** [reply]
Please use this tutorial (https://pnp.github.io/pnpcore/tutorials/azurefunctions/v4processisolatedapponly.html) as it shows how to setup certificate auth and use it from your local machine and from when deployed in Azure.

**Comment by necik11** [comment]
Finally I got one step further! 👍 Maybe its everything work, just for information:

I had almost everything right just cause i had many tenants i set wrong App ...so i didnt have permissions and got 401.

1. Register new Azure AD application -> 2. **Generate public certificate** -> 3. Add Certificate to new registred app -> 4. Use Thumbprint into X509 auth -> 5. Set application site.FullControlAll! 
2. I use loadCertificate method -  Environment.GetEnvironmentVariable and everything work like i wanted

And for be sure overview:
At my tenant i have public certificate with private key stored in keyvault - then make app service - there i set this certificate use from keyvault and call variable with url name: "CertificateFromKeyVault".

And after this i get my X509Certificate2 in code and for every another tenant where i will connect i will create new appregistration and set only public cert without password. ? And permission will be granted depend on application security and certification validation ? :)
🆗

**Comment by necik11** [comment]
Great, everything works! 👍 

In fact, there are more options for everything... but probably the most correct for 1:N app ->

- Is to create an app registration for multiple tenants - and share the link with other customers/tenants - to confirm that we can use the app with our needed permission.

- And for authenticating to the service, it is best to use auth with the certificate, same like you share [here](https://pnp.github.io/pnpcore/tutorials/azurefunctions/v4processisolatedapponly.html), and then set the appropriate permissions, either full control - or via sites.selected + API Graph/PnP Powershell set only on certain sites.

Thanks for the help and I hope it helps someone else too :) 😉
